pkg/database/database.go

```diff
@@ -0,0 +1,185 @@
+package database
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"github.com/appuio/appuio-cloud-reporting/pkg/db"
+	"github.com/jmoiron/sqlx"
+	"github.com/vshn/cloudscale-metrics-collector/pkg/categoriesmodel"
+	"github.com/vshn/cloudscale-metrics-collector/pkg/datetimesmodel"
+	"github.com/vshn/cloudscale-metrics-collector/pkg/discountsmodel"
+	"github.com/vshn/cloudscale-metrics-collector/pkg/factsmodel"
+	"github.com/vshn/cloudscale-metrics-collector/pkg/productsmodel"
+	"github.com/vshn/cloudscale-metrics-collector/pkg/queriesmodel"
+	"github.com/vshn/cloudscale-metrics-collector/pkg/tenantsmodel"
+	"net/url"
+	"strings"
+	"time"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+const (
+	sourceQueryStorage = "object-storage-storage"
+	provider           = "exoscale"
+	queryAndZone       = sourceQueryStorage + ":" + provider
+	defaultUnit        = "GBDay"
+)
+
+var (
+	product = db.Product{
+		Source: queryAndZone,
+		Target: sql.NullString{String: "1402", Valid: true},
+		Amount: 0.00066,
+		Unit:   "GBDay",
+		During: db.InfiniteRange(),
+	}
+	discount = db.Discount{
+		Source:   sourceQueryStorage,
+		Discount: 0,
+		During:   db.InfiniteRange(),
+	}
+	query = db.Query{
+		Name:        queryAndZone,
+		Description: "Object Storage - Storage (exoscale.com)",
+		Query:       "",
+		Unit:        "GBDay",
+		During:      db.InfiniteRange(),
+	}
+)
+
+// AggregatedBucket contains total used storage in an organization namespace
+type AggregatedBucket struct {
+	Organization string
+	// Storage in bytes
+	StorageUsed int64
+}
+
+// Database holds raw url of the postgresql database with the opened connection
+type Database struct {
+	Url        string
+	connection *sqlx.DB
+}
+
+// ValidateDatabaseURL validates environment variable dbUrlEnvVariable which holds the database URL
+func ValidateDatabaseURL(databaseURL string) (database *Database, err error) {
+	_, err = url.ParseRequestURI(databaseURL)
+	if err != nil {
+		return nil, fmt.Errorf("database url is not valid")
+	}
+	return &Database{Url: databaseURL}, nil
+}
```

Comment:
I have 2 issues here:
- It's not validation.
- This is not typical Go.

First, here you're only parsing an url, but not validating. It's not checked whether a user and password is given, for instance. Or the database and host. For example, `postgres:///db` is actually a parseable URL, but doesn't contain all the required settings.
```go
func TestParse(t *testing.T) {
	u, err := url.Parse("postgres:///db")
	assert.NoError(t, err)
	t.Logf("err: %v, scheme: %s, host: %s, path: %s, user: %s, port: %s", err, u.Scheme, u.Host, u.Path, u.User.String(), u.Port())
}
// output:
main_test.go:28: err: <nil>, scheme: postgres, host: , path: /db, user: , port: 
```

Second. This function does both "validation" and building a new instance. Either the name is wrongly used, or the validation should be its function, e.g. `func ValidateDatabaseURL(string) error`. (also, you should fully-parse the URL with `url.Parse()`, not just ParseRequestURI(), as that's more for parsing web requests). The advantage of having it separately is that it can be used as a Flag-level action in the cli library, to catch misconfiguration before attempting to build a DB instance. Currently, in `sos_command.go` it's used both to parse and then to create a new instance, but in a pre-action.

So, I would suggest doing something like this:
```go
func ValidateDatabaseURL(url string) error { 
	u, err := url.Parse(url)
	...
	// test if all required settings are present
}

// OpenConnection opens a connection to the postgres database
func OpenConnection(url string) (*Database, error) {
	connection, err := db.Openx(url)
	if err != nil {
		return nil, fmt.Errorf("cannot create a connection to the database: %w", err)
	}
	return &Database{Url: url, connection: connection}, nil
}
```

Reply:
I'm fine just validating the url and not return a `Database` object instance. Though I minimally validate the url with `ParseRequestURI`. I tried to pass some random values and `Parse` function would not return any error like `ParseRequestURI` did. I will add more validation options to the to check in depth if all parts from the uri are set.

Reply:
I'm also fine with not validating at all and just let lower-level functions (`db.Openx(url)`) fail the connection attempt if the URL is somehow missing properties.

Reply:
I will leave the validation as is and not add more code to it but will change the method return as per your suggestion.

Reply:
> Parsing is at some level a partial validation.

It's a syntactic validation, but not semantic. A syntactic validation doesn't help the user if the user and password is missing :) (that's the semantic one here).

Let's just get rid of parsing and validation completely. In this case I say either do it completely or not at all, but only parsing doesn't help anyone here (it doesn't help the user for missing properties and only creates more code to maintain).
Ofc you can decide, since I'm switching the team soon :)

Reply:
Let's remove it for now.